In [3]:
import pandas as pd

In [1]:
import supabase
print("Supabase töötab")


AttributeError: 'typing.Union' object has no attribute '__module__' and no __dict__ for setting new attributes

In [2]:
import pandas as pd

data = {
    'customer_id': [1001, 1002, 1003],
    'sale_date': ['2024-01-15', '2024-01-16', '2024-02-01'],
    'total_price': [89.99, 45.50, 120.00],
    'city': ['Tallinn', 'Tartu', 'Tallinn']
}

df = pd.DataFrame(data)

df.head()

,customer_id,sale_date,total_price,city
0,1001,2024-01-15,89.99,Tallinn
1,1002,2024-01-16,45.50,Tartu
2,1003,2024-02-01,120.00,Tallinn


In [3]:
df.head()

,customer_id,sale_date,total_price,city
0,1001,2024-01-15,89.99,Tallinn
1,1002,2024-01-16,45.50,Tartu
2,1003,2024-02-01,120.00,Tallinn


In [7]:
df.tail()

,customer_id,sale_date,total_price,city
0,1001,2024-01-15,89.99,Tallinn
1,1002,2024-01-16,45.50,Tartu
2,1003,2024-02-01,120.00,Tallinn


In [5]:
df.shape

(3, 4)

In [8]:
df.dtypes

customer_id      int64
sale_date          str
total_price    float64
city               str
dtype: object

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   customer_id  3 non-null      int64  
 1   sale_date    3 non-null      str    
 2   total_price  3 non-null      float64
 3   city         3 non-null      str    
dtypes: float64(1), int64(1), str(2)
memory usage: 228.0 bytes


In [10]:
df.describe()

,customer_id,total_price
count,3.0,3.000000
mean,1002.0,85.163333
std,1.0,37.483797
min,1001.0,45.500000
25%,1001.5,67.745000
50%,1002.0,89.990000
75%,1002.5,104.995000
max,1003.0,120.000000


# Harjutus 1A
## Ülesanne: Kopeeri järgmine kood Jupyter Notebook'i ja käivita see

## Vastused:
- Mitu rida ja veergu on DataFrame'is? - 20 rida ja 5 veergu
- Mis tüüpi on total_price veerg? - float64
- Mis on keskmine tellimuse summa (describe väljundist)? - 103.289500
- Mis on suurim üksik tellimus? - 230.000000


In [2]:
import pandas as pd

# Loo UrbanStyle näidisandmed (kui CSV puudub)
data = {
    'customer_id': [1001, 1002, 1003, 1001, 1002, 1004, 1003, 1001, 1005, 1004,
                    1002, 1003, 1005, 1001, 1006, 1004, 1002, 1007, 1003, 1005],
    'sale_date': ['2024-01-15', '2024-01-16', '2024-02-01', '2024-02-20', '2024-03-01',
                   '2024-03-05', '2024-03-15', '2024-04-10', '2024-04-12', '2024-04-20',
                   '2024-05-01', '2024-05-10', '2024-05-15', '2024-06-01', '2024-06-05',
                   '2024-06-10', '2024-06-20', '2024-07-01', '2024-07-05', '2024-07-10'],
    'total_price': [89.99, 45.50, 120.00, 67.30, 55.00, 210.00, 33.50, 145.00, 78.00, 92.00,
                     160.00, 44.00, 88.50, 230.00, 37.00, 175.00, 110.00, 65.00, 95.00, 125.00],
    'city': ['Tallinn', 'Tartu', 'Tallinn', 'Tallinn', 'Tartu', 'Pärnu', 'Tallinn', 'Tallinn',
             'Tartu', 'Pärnu', 'Tartu', 'Tallinn', 'Tartu', 'Tallinn', 'Pärnu', 'Pärnu',
             'Tartu', 'Tallinn', 'Tallinn', 'Tartu'],
    'product_category': ['Dresses', 'Tops', 'Denim', 'Accessories', 'Tops', 'Denim', 'Tops',
                         'Dresses', 'Denim', 'Accessories', 'Dresses', 'Tops', 'Denim',
                         'Dresses', 'Accessories', 'Denim', 'Tops', 'Accessories', 'Dresses', 'Denim']
}

df = pd.DataFrame(data)

# Uuri andmeid
print("=== SHAPE ===")
print(df.shape)

print("\n=== HEAD ===")
print(df.head())

print("\n=== INFO ===")
print(df.info())

print("\n=== DESCRIBE ===")
print(df.describe())

print("\n=== DTYPES ===")
print(df.dtypes)


=== SHAPE ===
(20, 5)

=== HEAD ===
   customer_id   sale_date  total_price     city product_category
0         1001  2024-01-15        89.99  Tallinn          Dresses
1         1002  2024-01-16        45.50    Tartu             Tops
2         1003  2024-02-01       120.00  Tallinn            Denim
3         1001  2024-02-20        67.30  Tallinn      Accessories
4         1002  2024-03-01        55.00    Tartu             Tops

=== INFO ===
<class 'pandas.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customer_id       20 non-null     int64  
 1   sale_date         20 non-null     str    
 2   total_price       20 non-null     float64
 3   city              20 non-null     str    
 4   product_category  20 non-null     str    
dtypes: float64(1), int64(1), str(3)
memory usage: 932.0 bytes
None

=== DESCRIBE ===
       customer_id  total_price
count     20.000

In [13]:
print("\n=== DESCRIBE ===")
print(df.describe())

print("\n=== DTYPES ===")
print(df.dtypes)


=== DESCRIBE ===
       customer_id  total_price
count     20.00000    20.000000
mean    1003.20000   103.289500
std        1.76516    56.559621
min     1001.00000    33.500000
25%     1002.00000    62.500000
50%     1003.00000    90.995000
75%     1004.25000   130.000000
max     1007.00000   230.000000

=== DTYPES ===
customer_id           int64
sale_date               str
total_price         float64
city                    str
product_category        str
dtype: object


In [14]:
# 1. Mitu unikaalset klienti on andmestikus?
print("Unikaalseid kliente:", df['customer_id'].nunique())

# 2. Millistes linnades asuvad meie kliendid?
print("Linnad:", df['city'].unique())

# 3. Mitu tellimust on igast linnast?
print("Tellimused linniti:\n", df['city'].value_counts())

# 4. Mis on kogu käive?
print("Kogukäive:", df['total_price'].sum())

Unikaalseid kliente: 7
Linnad: <StringArray>
['Tallinn', 'Tartu', 'Pärnu']
Length: 3, dtype: str
Tellimused linniti:
 city
Tallinn    9
Tartu      7
Pärnu      4
Name: count, dtype: int64
Kogukäive: 2065.79


- Millisest linnast tuleb enim tellimusi? - Tallinn, 9
- Kas see on üllatav, arvestades UrbanStyle'i profiili? Ei ole, suurim pood asub seal
- Kuidas sa sama küsimuse SQL-is vastaksid?
```sql 
SELECT city, COUNT(*) AS tellimuste_arv
FROM sales
GROUP BY city
ORDER BY tellimuste_arv DESC;


## Harjutus 1C

# Ülesanne: Vali DataFrame'ist üks veerg, mida pole veel uuritud (nt product_category). Kirjuta 3-5 rida koodi, mis annavad sulle selle veeru kohta ülevaate.

# Vastused>
## Tähelepanekud

- Kõige populaarsem kategooria oli Dresses.
- Accessories kategoorias oli vähem tellimusi.
- See aitab mõista, millised tooted müüvad kõige paremini.

In [19]:
print("Kategooriad:", df['product_category'].unique())

print("Mitu kategooriat:", df['product_category'].nunique())

print("Tellimused kategooriate kaupa:")
print(df['product_category'].value_counts())



Kategooriad: <StringArray>
['Dresses', 'Tops', 'Denim', 'Accessories']
Length: 4, dtype: str
Mitu kategooriat: 4
Tellimused kategooriate kaupa:
product_category
Denim          6
Dresses        5
Tops           5
Accessories    4
Name: count, dtype: int64


In [20]:
df[df['total_price'] > 100]

,customer_id,sale_date,total_price,city,product_category
2,1003,2024-02-01,120.0,Tallinn,Denim
5,1004,2024-03-05,210.0,Pärnu,Denim
7,1001,2024-04-10,145.0,Tallinn,Dresses
10,1002,2024-05-01,160.0,Tartu,Dresses
13,1001,2024-06-01,230.0,Tallinn,Dresses
15,1004,2024-06-10,175.0,Pärnu,Denim
16,1002,2024-06-20,110.0,Tartu,Tops
19,1005,2024-07-10,125.0,Tartu,Denim


### Mõtlemisülesanne:

Kuidas sa arvad, mis on pandas'i vaste neile SQL lausetele?

| SQL | pandas |
|---|---|
| WHERE city = 'Tallinn' | `df[df['city'] == 'Tallinn']` |
| WHERE totalprice > 100 AND city = 'Tallinn' | `df[(df['total_price'] > 100) & (df['city'] == 'Tallinn')]` |
| GROUP BY storelocation ... SUM(totalprice) | `df.groupby('store_location')['total_price'].sum()` |
 



## Samm 1: Boolean indexing (filtreerimine)
SQL-i WHERE lause pandas'i vastes:



In [ ]:

# Lihtne filter
tallinna_tellimused = df[df['city'] == 'Tallinn']

# Mitu tingimust (AND = &, OR = |)
suured_tallinn = df[(df['total_price'] > 100) & (df['city'] == 'Tallinn')]

# NOT (tilde ~)
mitte_tallinn = df[~(df['city'] == 'Tallinn')]

In [ ]:
## Samm 2: groupby() — SQL GROUP BY pandas'is
# SQL: SELECT city, SUM(total_price) FROM sales GROUP BY city;

# Pandas:
df.groupby('city')['total_price'].sum()

# Mitu agregaatfunktsiooni korraga
df.groupby('city')['total_price'].agg(['sum', 'mean', 'count'])

# Mitu grupeerimise veergu
df.groupby(['city', 'product_category'])['total_price'].sum()


city     product_category
Pärnu    Accessories         129.00
         Denim               385.00
Tallinn  Accessories         132.30
         Denim               120.00
         Dresses             559.99
         Tops                 77.50
Tartu    Denim               291.50
         Dresses             160.00
         Tops                210.50
Name: total_price, dtype: float64

In [25]:
## Samm 3: sort_values() — SQL ORDER BY pandas'is
# Sorteeri käive kahanevalt
df.sort_values('total_price', ascending=False)

# Sorteeri mitme veeru järgi
df.sort_values(['city', 'total_price'], ascending=[True, False])


,customer_id,sale_date,total_price,city,product_category
5,1004,2024-03-05,210.00,Pärnu,Denim
15,1004,2024-06-10,175.00,Pärnu,Denim
9,1004,2024-04-20,92.00,Pärnu,Accessories
14,1006,2024-06-05,37.00,Pärnu,Accessories
13,1001,2024-06-01,230.00,Tallinn,Dresses
7,1001,2024-04-10,145.00,Tallinn,Dresses
2,1003,2024-02-01,120.00,Tallinn,Denim
18,1003,2024-07-05,95.00,Tallinn,Dresses
0,1001,2024-01-15,89.99,Tallinn,Dresses
3,1001,2024-02-20,67.30,Tallinn,Accessories


## Samm 4: merge() — SQL JOIN pandas'is
# See on üks olulisemaid operatsioone. Mõtle tagasi Week 3 JOIN-idele:
# SQL: SELECT * FROM sales s JOIN customers c ON s.customer_id = c.customer_id
# Pandas:
merged = pd.merge(df_sales, df_customers, on='customer_id', how='left')
SQL JOIN
pandas merge how=
INNER JOIN
how='inner' (vaikimisi)
LEFT JOIN
how='left'
RIGHT JOIN
how='right'
FULL OUTER JOIN
how='outer'

In [27]:
## Samm 5: Uute veergude loomine ja apply()
# Uus veerg arvutusega
df['discount'] = df['total_price'] * 0.1

# Tingimuslik veerg (nagu SQL CASE WHEN)
df['segment'] = df['total_price'].apply(
    lambda x: 'Suur' if x > 100 else 'Väike'
)

# Kuupäeva teisendamine
df['sale_date'] = pd.to_datetime(df['sale_date'])
df['month'] = df['sale_date'].dt.month
df['year'] = df['sale_date'].dt.year


In [ ]:
## Harjutus 2A

# Ülesanne: Kopeeri ja käivita järgmine kood, mis replitseerib Week 1-4 SQL analüüse pandas'iga.
import pandas as pd

# Kasuta Osa 1 DataFrame'i (df) või laadi uuesti
# df = pd.read_csv('urbanstyle_sales.csv')

# 1. Filtreerimine (WHERE)
print("=== Tallinna tellimused ===")
tallinn = df[df['city'] == 'Tallinn']
print(f"Kokku: {len(tallinn)} tellimust, käive: {tallinn['total_price'].sum():.2f} EUR")

# 2. Grupeerimie (GROUP BY)
print("\n=== Käive linniti ===")
city_revenue = df.groupby('city')['total_price'].agg(['sum', 'mean', 'count'])
city_revenue.columns = ['kogukäive', 'keskmine', 'tellimusi']
print(city_revenue.sort_values('kogukäive', ascending=False))

# 3. Uue veeru loomine (CASE WHEN)
df['order_size'] = df['total_price'].apply(
    lambda x: 'Suur (100+)' if x >= 100 else 'Väike (<100)'
)
print("\n=== Tellimused suuruse järgi ===")
print(df['order_size'].value_counts())



=== Tallinna tellimused ===
Kokku: 9 tellimust, käive: 889.79 EUR

=== Käive linniti ===
         kogukäive    keskmine  tellimusi
city                                     
Tallinn     889.79   98.865556          9
Tartu       662.00   94.571429          7
Pärnu       514.00  128.500000          4

=== Tellimused suuruse järgi ===
order_size
Väike (<100)    12
Suur (100+)      8
Name: count, dtype: int64


# Vastused:
- Mis on Tallinna kogukäive? - 889,79
- Milline linn on keskmise tellimuse poolest esikohal? - Pärnu, 128,5
- Mitu "suurt" tellimust on? - 8


## Harjutus 2B
# Ülesanne: Marko tahab teada iga kliendi ostukäitumist. Kirjuta kood, mis arvutab iga kliendi kohta: mitu korda ta ostis, kui palju ta kokku kulutas ja mis oli tema keskmine tellimus.
# Vihjed:
groupby('customer_id') + .agg({'total_price': ['sum', 'mean', 'count']})
Sorteeri kogukulutuse järgi kahanevalt
Lisa veerg 'vip_status', kus kogukulutus > 200 on 'VIP'


In [11]:

customer_summary = df.groupby('customer_id')['total_price'].agg(
    kogukulutus='sum',
    keskmine='mean',
    tellimusi='count'
).reset_index()

# Sorteeri
customer_summary = customer_summary.sort_values('kogukulutus', ascending=False)

# Lisa VIP staatus
customer_summary['vip_status'] = customer_summary['kogukulutus'].apply(
    lambda x: 'VIP' if x > 200 else 'Tavaline'
)

print(customer_summary)

   customer_id  kogukulutus    keskmine  tellimusi vip_status
0         1001       532.29  133.072500          4        VIP
3         1004       477.00  159.000000          3        VIP
1         1002       370.50   92.625000          4        VIP
2         1003       292.50   73.125000          4        VIP
4         1005       291.50   97.166667          3        VIP
6         1007        65.00   65.000000          1   Tavaline
5         1006        37.00   37.000000          1   Tavaline


## Vastused:
- Kes on kõige väärtuslikum klient? - customer_id 1001
- Mitu VIP-klienti on? 5
- Miks see info on Markole kasulik? - Näeb täpselt, kes ja kui palju kulutab


In [31]:
## Harjutus 2C: Rakendus — merge ja analüüs (10 min)

# Ülesanne: Loo teine DataFrame kliendiandmetega ja ühenda see müügiandmetega (merge). Analüüsi tulemust.

# Kliendiandmed
customers = pd.DataFrame({
    'customer_id': [1001, 1002, 1003, 1004, 1005, 1006, 1007],
    'first_name': ['Jüri', 'Kati', 'Maris', 'Peeter', 'Liina', 'Andres', 'Tiina'],
    'email': ['juri@mail.ee', 'kati@mail.ee', 'maris@mail.ee', 'peeter@mail.ee',
              'liina@mail.ee', 'andres@mail.ee', 'tiina@mail.ee']
})

# Ühenda müügiandmetega
merged = pd.merge(df, customers, on='customer_id', how='left')

# Nüüd saad näidata kliendi NIME, mitte ainult ID-d
# Kirjuta kood, mis näitab TOP 5 klienti nime ja kogukulutuse järgi

# Kontroll pärast merge
print("merged shape:", merged.shape)

# TOP 5 klienti kogukulutuse järgi
top_clients = merged.groupby('first_name')['total_price'].sum().sort_values(ascending=False).head(5)

print("\nTOP 5 kliendid:")
print(top_clients)



merged shape: (20, 12)

TOP 5 kliendid:
first_name
Jüri      532.29
Peeter    477.00
Kati      370.50
Maris     292.50
Liina     291.50
Name: total_price, dtype: float64


## Kontrolltabel:
- merge() töötab korrektselt (kontrollisid shape enne ja pärast) -jah, ühendatud tabel sisaldab 20 rida ja 12 veergu
- Ma oskan seletada, miks kasutasin how='left' (mitte how='inner') - Kasutasin `how='left'`, et kõik müügiandmed säiliksid ka siis, kui mõnel kliendil puuduks lisainfo.
- Tulemus näitab kliendi nime, mitte ainult ID-d - Tulemus näitab nüüd kliendi nime, mitte ainult ID-d.
- Kõige suurema kogukulutusega klient oli Jüri (532.29 EUR).
- TOP 5 kliendid võivad olla UrbanStyle'i kõige väärtuslikumad püsikliendid.


##  Plotly Express samm-sammult 
### Samm 1: Plotly import ja põhiloogika
import plotly.express as px
Plotly Express (lühend px) on Plotly kõrgtaseme API — see on loodud kiireks visualiseerimiseks. Iga graafik järgib sama mustrit:
fig = px.diagrammi_tüüp(
    data_frame=df,         # Andmed
    x='x_veerg',           # X-telg
    y='y_veerg',           # Y-telg
    color='kategooria',    # Värvide järgi jaotus (valikuline)
    title='Pealkiri'       # Graafiku pealkiri
)
fig.show()                  # Näita graafikut


In [1]:
import plotly.express as px


In [1]:
import pandas as pd
import plotly.express as px

In [3]:
df = pd.DataFrame(data)

## Samm 2: Põhidiagrammid

In [4]:
# Tulpdiagramm (bar chart)
# Käive linniti
city_data = df.groupby('city')['total_price'].sum().reset_index()
fig = px.bar(
    city_data,
    x='city',
    y='total_price',
    title='UrbanStyle käive linnade kaupa',
    labels={'city': 'Linn', 'total_price': 'Käive (EUR)'}
)
fig.show()




In [15]:
## Joondiagramm (line chart):
# Käive kuude kaupa
df['sale_date'] = pd.to_datetime(df['sale_date'])
monthly = df.groupby(df['sale_date'].dt.to_period('M'))['total_price'].sum().reset_index()
monthly['sale_date'] = monthly['sale_date'].astype(str)

fig = px.line(
    monthly,
    x='sale_date',
    y='total_price',
    title='UrbanStyle kuukäive trend',
    labels={'sale_date': 'Kuu', 'total_price': 'Käive (EUR)'},
    markers=True
)
fig.show()


In [20]:
## Hajuvusdiagramm (scatter plot):
# Kliendi ostukäitumine: tellimuste arv vs kogukulutus
fig = px.scatter(
    customer_summary,
    x='tellimusi',
    y='kogukulutus',
    color='vip_status',
    title='Kliendid: ostusagedus vs kogukulutus',
    labels={'tellimusi': 'Tellimuste arv', 'kogukulutus': 'Kogukulutus (EUR)'}
)
fig.show()


In [21]:
## Sektordiagramm (pie chart):
# Käive osakaal kategooriate kaupa
cat_data = df.groupby('product_category')['total_price'].sum().reset_index()
fig = px.pie(
    cat_data,
    values='total_price',
    names='product_category',
    title='Käive jaotus tootekategooriate kaupa'
)
fig.show()

In [5]:
## Samm 3: Kohandamine
# Värvide kohandamine
fig.update_layout(
    plot_bgcolor='white',
    font=dict(family='Calibri', size=14),
    title_font_size=18
)

# Telgede kohandamine
fig.update_xaxes(title_text='Linn', tickangle=45)
fig.update_yaxes(title_text='Käive (EUR)')

# Salvestamine HTML-ina
fig.write_html('urbanstyle_chart.html')

# Salvestamine pildina (vajab kaleido paketti)
# pip install kaleido
fig.write_image('urbanstyle_chart.png')


## Harjutus 3A
# Ülesanne: Kopeeri ja käivita järgmine kood. Seejärel muuda vähemalt ühte asja (pealkiri, värvid, diagrammi tüüp).

In [6]:
import pandas as pd
import plotly.express as px

# Kasuta olemasolevat df-i või loo uuesti (vt Osa 1A)

# Käive kategooriate kaupa
cat_revenue = df.groupby('product_category')['total_price'].sum().reset_index()
cat_revenue = cat_revenue.sort_values('total_price', ascending=True)

fig = px.bar(
    cat_revenue,
    x='total_price',
    y='product_category',
    orientation='h',
    title='UrbanStyle: Käive tootekategooriate kaupa',
    labels={
        'total_price': 'Käive (EUR)',
        'product_category': 'Kategooria'
    },
    color='product_category'
)
fig.show()

In [7]:
fig = px.bar(
    cat_revenue,
    x='total_price',
    y='product_category',
    orientation='h',
    title='UrbanStyle: Kõige edukamad tootekategooriad',
    labels={
        'total_price': 'Käive (EUR)',
        'product_category': 'Kategooria'
    },
    color='product_category',
    text='total_price'
)

fig.update_layout(showlegend=False)

fig.show()

In [8]:
fig = px.bar(
    cat_revenue,
    x='total_price',
    y='product_category',
    orientation='h',
    title='UrbanStyle: Kõige edukamad tootekategooriad',
    color='product_category',
    text='total_price',
    color_discrete_sequence=['pink', 'purple', 'orange', 'blue']
)

fig.show()

## Harjutus 3B
# Ülesanne: Loo kaks graafikut UrbanStyle andmetest, mis vastavad Marko küsimustele.
- Graafik 1: Kuukäive trend (joondiagramm)
X-telg: kuu
Y-telg: käive
Lisa markers=True punktide kuvamiseks
- Graafik 2: Klientide jaotus VIP-staatuse järgi (sektordiagramm)
Kasuta Osa 2 customer_summary DataFrame'i
Näita VIP vs mitte-VIP klientide arvu

# Vastused:
- Milline kuu oli kõige kasumlikum? - Juuni, 2024
- Mis protsent klientidest on VIP? 71,4%
- Millist graafikut näitaksid sa Markole esimesena? Miks? Kuukäive trendi, siis Marko näeb, mis kuudel on müüki ja kus vajaks rohkem tegeleda kampaaniatega nt



In [9]:
# Kuukäive trend

df['sale_date'] = pd.to_datetime(df['sale_date'])

monthly = df.groupby(df['sale_date'].dt.to_period('M'))['total_price'].sum().reset_index()
monthly['sale_date'] = monthly['sale_date'].astype(str)

fig = px.line(
    monthly,
    x='sale_date',
    y='total_price',
    title='UrbanStyle: Kuukäive trend',
    labels={'sale_date': 'Kuu', 'total_price': 'Käive (EUR)'},
    markers=True
)

fig.show()

In [12]:
# Klientide jaotus VIP-staatuse järgi

vip_counts = customer_summary['vip_status'].value_counts().reset_index()

fig = px.pie(
    vip_counts,
    values='count',
    names='vip_status',
    title='UrbanStyle: Kliendid VIP-staatuse järgi'
)

fig.show()

# Harjutus 3C - andmelugu graafikuga
## Ülesanne: Mõtle äritegevusele, mida sa tahaksid Markole näidata. Loo üks graafik, mis räägib loo.
Knaflic Ch 7 soovitused:
Seadistus: "UrbanStyle'il on arv] klienti..."
Konflikt: "Aga mõned kliendid..."
Lahendus: "Andmed näitavad, et..."

## Kontrolltabel:
- Graafik on korrektselt sildistatud (pealkiri, teljed)
- Graafik vastab konkreetsele äriküsimusele
- Ma oskan seletada, mida graafik näitab (30 sekundit)

### Seadistus
UrbanStyle'il on 7 klienti selles näidisandmestikus.

### Konflikt
Aga kõik kliendid ei panusta käibesse võrdselt — osa kliente kulutab oluliselt rohkem kui teised.

### Lahendus
Andmed näitavad, et VIP-kliendid moodustavad kõige väärtuslikuma segmendi. Marko võiks neile pakkuda eraldi lojaalsuskampaaniat.



In [13]:
# Andmelugu VIP-klientidest
customer_summary = df.groupby('customer_id')['total_price'].agg(
    kogukulutus='sum',
    keskmine='mean',
    tellimusi='count'
).reset_index()

customer_summary['vip_status'] = customer_summary['kogukulutus'].apply(
    lambda x: 'VIP' if x > 200 else 'Tavaline'
)

fig = px.bar(
    customer_summary.sort_values('kogukulutus', ascending=False),
    x='customer_id',
    y='kogukulutus',
    color='vip_status',
    title='UrbanStyle: Kliendid kogukulutuse järgi',
    labels={
        'customer_id': 'Kliendi ID',
        'kogukulutus': 'Kogukulutus (EUR)',
        'vip_status': 'VIP staatus'
    },
    text='kogukulutus'
)

fig.show()

# Concepts: RFM raamistik

RFM on kliendisegmenteerimise klassika, mida kasutavad e-kaubandusettevõtted üle maailma.

| Dimensioon | Küsimus | Arvutus |
|---|---|---|
| **Recency** | Kui hiljuti klient ostis? | Päevi viimasest ostust |
| **Frequency** | Kui tihti klient ostab? | Ostude arv perioodi jooksul |
| **Monetary** | Kui palju klient kulutab? | Kogukulutus perioodi jooksul |

---

## Skoorid (1–5)

- **Recency:** `R=5` (ostis hiljuti) → `R=1` (pole ammu ostnud)
- **Frequency:** `F=5` (ostab tihti) → `F=1` (ostis ainult korra)
- **Monetary:** `M=5` (kulutab palju) → `M=1` (kulutab vähe)

---

## RFM koguskoor

```text
RFM koguskoor = R + F + M
```

Vahemik: **3–15**

---

## Segmendid

| Skoor | Segment | Markole tähendab |
|---|---|---|
| **13–15** | VIP Champions | VIP-programm, eriline kohtlemine |
| **10–12** | Loyal Customers | Lojaalsusprogramm |
| **7–9** | Potential Loyalists | Veel üks ost ja nad on VIP! |
| **4–6** | At Risk | "Me igatseme teid!" kampaania |
| **3** | Lost | Kadunud — kas saame tagasi? |

# Integreeriv RFM harjutus 
## Ülesanne: Vii läbi lihtsustatud RFM analüüs, kasutades kõiki kolme Osa teadmisi.


In [14]:
# Samm 1 - andmete laadimine

import pandas as pd
import plotly.express as px

# Kasuta olemasolevat df-i või loo uuesti (vt Osa 1A)
# Veendu, et sale_date on datetime tüüpi
df['sale_date'] = pd.to_datetime(df['sale_date'])

# Viitekuupäev (viimase tellimuse järgmine päev)
today = pd.to_datetime('2024-08-01')
print("Andmestiku periood:", df['sale_date'].min(), "kuni", df['sale_date'].max())
print("Viitekuupäev:", today)


Andmestiku periood: 2024-01-15 00:00:00 kuni 2024-07-10 00:00:00
Viitekuupäev: 2024-08-01 00:00:00


In [15]:
# Samm 2 - RFM arvutamine

# Recency: päevi viimasest ostust
recency = df.groupby('customer_id')['sale_date'].max().reset_index()
recency.columns = ['customer_id', 'last_purchase']
recency['recency_days'] = (today - recency['last_purchase']).dt.days

# Frequency: ostude arv
frequency = df.groupby('customer_id').size().reset_index(name='frequency')

# Monetary: kogukulutus
monetary = df.groupby('customer_id')['total_price'].sum().reset_index()
monetary.columns = ['customer_id', 'monetary']

# Liida kokku
rfm = recency[['customer_id', 'recency_days']].merge(
    frequency, on='customer_id'
).merge(
    monetary, on='customer_id'
)

print(rfm)


   customer_id  recency_days  frequency  monetary
0         1001            61          4    532.29
1         1002            42          4    370.50
2         1003            27          4    292.50
3         1004            52          3    477.00
4         1005            22          3    291.50
5         1006            57          1     37.00
6         1007            31          1     65.00


In [16]:
# Samm 3 - skooride määramine

# Lihtsustatud skooride määramine (täida lüngad!)
# Recency: väiksem = parem (hiljuti ostis), seega pööra ümber
rfm['R_score'] = pd.qcut(rfm['recency_days'], q=3, labels=[3, 2, 1]).astype(int)

# Frequency: suurem = parem
rfm['F_score'] = pd.qcut(
    rfm['frequency'].rank(method='first'), q=3, labels=[1, 2, 3]
).astype(int)

# Monetary: suurem = parem
rfm['M_score'] = pd.qcut(rfm['monetary'], q=3, labels=[1, 2, 3]).astype(int)

# Koguskoor
rfm['RFM_score'] = rfm['R_score'] + rfm['F_score'] + rfm['M_score']

# Segmenteerimine
def assign_segment(score):
    if score >= 8:
        return 'VIP Champions'
    elif score >= 6:
        return 'Loyal Customers'
    elif score >= 4:
        return 'Potential Loyalists'
    else:
        return 'At Risk'

rfm['segment'] = rfm['RFM_score'].apply(assign_segment)
print(rfm.sort_values('RFM_score', ascending=False))


   customer_id  recency_days  frequency  monetary  R_score  F_score  M_score  \
2         1003            27          4    292.50        3        3        2   
1         1002            42          4    370.50        2        3        2   
0         1001            61          4    532.29        1        2        3   
3         1004            52          3    477.00        2        1        3   
4         1005            22          3    291.50        3        2        1   
6         1007            31          1     65.00        3        1        1   
5         1006            57          1     37.00        1        1        1   

   RFM_score              segment  
2          8        VIP Champions  
1          7      Loyal Customers  
0          6      Loyal Customers  
3          6      Loyal Customers  
4          6      Loyal Customers  
6          5  Potential Loyalists  
5          3              At Risk  


In [1]:
# Samm 4 - visualiseerimine

# Graafik 1: Segmentide jaotus
segment_counts = rfm['segment'].value_counts().reset_index()
fig1 = px.bar(
    segment_counts,
    x='segment',
    y='count',
    title='UrbanStyle: Kliendisegmentide jaotus (RFM)',
    labels={'segment': 'Segment', 'count': 'Klientide arv'},
    color='segment'
)
fig1.show()

# Graafik 2: Hajuvusdiagramm — Recency vs Monetary
fig2 = px.scatter(
    rfm,
    x='recency_days',
    y='monetary',
    color='segment',
    size='frequency',
    hover_data=['customer_id'],
    title='UrbanStyle: Recency vs Monetary (RFM)',
    labels={
        'recency_days': 'Päevi viimasest ostust',
        'monetary': 'Kogukulutus (EUR)'
    }
)
fig2.show()


NameError: name 'rfm' is not defined

# Küsimused:
- Mitu VIP Champions klienti leidsid? 1 
- Kui suur on VIP-klientide osakaal kogutuludest? 14,2%
- Millist soovitust annaksid Markole At Risk segmendi kohta? At Risk kliendid pole ammu ostnud ja nende kulutus on madal. Marko võiks teha personaalse tagasivõitmise kampaania või pakkuda allahindlust järgmisele ostule.


In [19]:
# mitu % on osakaal VIP-del kogutulust?

vip_revenue = rfm[rfm['segment'] == 'VIP Champions']['monetary'].sum()

total_revenue = rfm['monetary'].sum()

vip_percentage = (vip_revenue / total_revenue) * 100

print(f"VIP Champions osakaal kogutuludest: {vip_percentage:.1f}%")

VIP Champions osakaal kogutuludest: 14.2%
